### DISCUSSION TITLE
ASN Roles

### DISCUSSION THREAD

**Discussion by ryanmerolle** [netbox-community/netbox#8447] [Ideas]
Anyone have a use for ASN roles similar to or just extending Prefix/VLAN roles?

The alternative would be to setup a custom field for this, but I wanted to see how people thought about this just incase people thought this would be a good feature.

If the current Prefix/VLAN role was just extended to include ASNs, not much would need to be changed with the existing model since it is already called `ipam.role`.

**Comment by DanSheps** [comment]
What is your use case?  :D

**Reply by ryanmerolle** [reply]
Fair enough. I’m trying to walk through the ASNs and how I should approach or see if a feature request would be useful.

Let me step back.  

In previous NetBox versions, before the ASN model, I would leverage the single site ASN as the base ASN for the site.  From there I would use:
- The base ASN for the spines
- The spine ASN + 1 for all the l3leafs
- The spine ASN + 2 for all the border leafs (wan routers and where firewalls hang off of)
- The base offset of ASNs continue from here for other devices in the site

Any public ASNs were assigned to a region using config context and used for use in external peerings for that region.

I’m really trying to rethink an approach to this all with ASNs now in their own model:
- If ASNs were to be able to be assigned to IPAM roles I could use the role logic to assign ASNs to devices and peerings (local as and remote as).
-  Devices could be associated ASNs, but that might open a bigger can of worms with modeling peering and such.
- Keep the current method I mentioned with only one ASN assigned per site


I’m really thinking out loud here if anyone has any thoughts or could even use my previous approach in their own workflows.

**Reply by DanSheps** [reply]
Since you can assign multiple ASN's per site, I would probably use tagging instead of a role.  I don't think role is appropriate here and tagging would be more what you are looking for.

**Reply by ryanmerolle** [reply]
I would move towards a custom field drop down before a tag given I've seen user get into issues around tags with regards to overall complexity and it being free from.

Another option could be to leverage the upcoming #7006 in order to tie a device role to an ASN.  Edge Leafs would have 2 ASNs, but any of the rfc6996 / rfc5398 RIR ASN would be used for external peering vs the internal RIR associated record for the system AS.

Examples:
```
- ASN: 999
  RIR: ARIN
  custom_object_role: edge_leaf
- ASN: 64512 
  RIR: rfc6996 
  custom_object_role: edge_leaf
```
